File: rxui-sample-java/src/test/java/com/artemzin/rxui/sample/java/TestMainView.java
package com.artemzin.rxui.sample.java;

import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;
import io.reactivex.functions.Function;
import io.reactivex.subjects.PublishSubject;

import static com.artemzin.rxui.test.TestRxUi.testUi;
import static org.mockito.Mockito.mock;

class TestMainView implements MainView {

    final PublishSubject<String> login = PublishSubject.create();
    final PublishSubject<String> password = PublishSubject.create();
    final PublishSubject<Object> signInClicks = PublishSubject.create();

    Consumer<Object> signInEnable = mock(Consumer.class);
    Consumer<Object> signInDisable = mock(Consumer.class);
    Consumer<AuthService.Success> signInSuccess = mock(Consumer.class);
    Consumer<AuthService.Failure> signInFailure = mock(Consumer.class);

    @Override
    public Observable<String> login() {
        return login;
    }

    @Override
    public Observable<String> password() {
        return password;
    }

    @Override
    public Observable<Object> signInClicks() {
        return signInClicks;
    }

    @Override
    public Function<Observable<Object>, Disposable> singInEnable() {
        return testUi(signInEnable); // No need for imitation of Main Thread scheduler!
    }

    @Override
    public Function<Observable<Object>, Disposable> singInDisable() {
        return testUi(signInDisable);
    }

    @Override
    public Function<Observable<AuthService.Success>, Disposable> signInSuccess() {
        return testUi(signInSuccess);
    }

    @Override
    public Function<Observable<AuthService.Failure>, Disposable> signInFailure() {
        return testUi(signInFailure);
    }
}

File: rxui-sample-kotlin/src/test/kotlin/com/artemzin/rxui/sample/kotlin/TestMainView.kt
package com.artemzin.rxui.sample.kotlin

import com.artemzin.rxui.sample.kotlin.AuthService.Response.Failure
import com.artemzin.rxui.sample.kotlin.AuthService.Response.Success
import com.artemzin.rxui.test.TestRxUi.testUi
import io.reactivex.functions.Consumer
import io.reactivex.subjects.PublishSubject
import org.mockito.Mockito

class TestMainView : MainView {

    val signInEnableAction = Mockito.mock(Consumer::class.java) as Consumer<Unit>
    val signInDisableAction = Mockito.mock(Consumer::class.java) as Consumer<Unit>
    val signInSuccessAction = Mockito.mock(Consumer::class.java) as Consumer<Success>
    val signInFailureAction = Mockito.mock(Consumer::class.java) as Consumer<Failure>

    // Produces.
    override val login = PublishSubject.create<String>()
    override val password = PublishSubject.create<String>()
    override val signInClicks = PublishSubject.create<Unit>()

    // Consumes.
    override val signInEnable = testUi { enable: Unit -> signInEnableAction.accept(Unit) }
    override val signInDisable = testUi { disable: Unit -> signInDisableAction.accept(Unit) }
    override val signInSuccess = testUi { success: Success -> signInSuccessAction.accept(success) }
    override val signInFailure = testUi { failure: Failure -> signInFailureAction.accept(failure) }
}
